File: visualization/pvrp.m
%% Period Vehicle Routing Problem (PVRP)
NUM_CUSTOMERS = 20;
NUM_VEHICLES = 4;
VEHICLE_CAPACITY = 10;
MAX_ITER = 10000;
INITIAL_TEMP = 100.0;
COOLING_RATE = 0.003;
PERIOD_LENGTH = 7;
customers = struct('id', [], 'demand', []);
for i = 1:NUM_CUSTOMERS
    customers(i).id = i;
    customers(i).demand = randi([1, 5]); 
end
calculate_distance = @(c1, c2) abs(c1.id - c2.id);
function route_distance = calculate_route_distance(route, customers, calculate_distance)
    route_distance = 0;
    for i = 1:length(route)-1
        route_distance = route_distance + calculate_distance(customers(route(i)), customers(route(i+1)));
    end
    route_distance = route_distance + calculate_distance(customers(route(end)), customers(route(1)));
end
function initial_solution = generate_initial_solution(customers, NUM_VEHICLES, VEHICLE_CAPACITY, calculate_distance)
    initial_solution = cell(1, NUM_VEHICLES);
    visited = false(1, length(customers));
    visited(1) = true; 
    current_vehicle = 1;
    current_capacity = 0;
    current_customer = 1; 
    while true
        nearest_customer = -1;
        min_distance = inf;
        for i = 2:length(customers)
            if ~visited(i) && customers(i).demand + current_capacity <= VEHICLE_CAPACITY
                dist = calculate_distance(customers(current_customer), customers(i));
                if dist < min_distance
                    min_distance = dist;
                    nearest_customer = i;
                end
            end
        end
        if nearest_customer == -1
            current_vehicle = current_vehicle + 1;
            if current_vehicle > NUM_VEHICLES
                break;
            end
            current_capacity = 0;
            current_customer = 1; 
        else
            initial_solution{current_vehicle} = [initial_solution{current_vehicle}, nearest_customer];
            visited(nearest_customer) = true;
            current_capacity = current_capacity + customers(nearest_customer).demand;
            current_customer = nearest_customer;
        end
    end
end
function best_solution = simulated_annealing(customers, NUM_VEHICLES, MAX_ITER, INITIAL_TEMP, COOLING_RATE, VEHICLE_CAPACITY, calculate_distance, calculate_route_distance)
    current_solution = generate_initial_solution(customers, NUM_VEHICLES, VEHICLE_CAPACITY, calculate_distance);
    best_solution = current_solution;
    temperature = INITIAL_TEMP;
    for iter = 1:MAX_ITER
        new_solution = current_solution;
        vehicle1 = randi(NUM_VEHICLES);
        vehicle2 = randi(NUM_VEHICLES);
        if vehicle1 ~= vehicle2 && ~isempty(new_solution{vehicle1}) && ~isempty(new_solution{vehicle2})
            rand_index1 = randi(length(new_solution{vehicle1}));
            rand_index2 = randi(length(new_solution{vehicle2}));
            temp = new_solution{vehicle1}(rand_index1);
            new_solution{vehicle1}(rand_index1) = new_solution{vehicle2}(rand_index2);
            new_solution{vehicle2}(rand_index2) = temp;
        end
        current_cost = 0;
        new_cost = 0;
        for v = 1:NUM_VEHICLES
            current_cost = current_cost + calculate_route_distance(current_solution{v}, customers, calculate_distance);
            new_cost = new_cost + calculate_route_distance(new_solution{v}, customers, calculate_distance);
        end
        if new_cost < current_cost
            current_solution = new_solution;
        else
            acceptance_probability = exp((current_cost - new_cost) / temperature);
            if rand < acceptance_probability
                current_solution = new_solution;
            end
        end
        if calculate_route_distance(current_solution{1}, customers, calculate_distance) < calculate_route_distance(best_solution{1}, customers, calculate_distance)
            best_solution = current_solution;
        end
        temperature = temperature * (1 - COOLING_RATE);
    end
end
function plot_routes(customers, solution, NUM_VEHICLES)
    figure;
    hold on;
    title('PVRP');
    xlabel('X Coordinate');
    ylabel('Y Coordinate');
    plot(customers(1).id, 0, 'ko', 'MarkerSize', 10, 'LineWidth', 2);
    text(customers(1).id, 0, 'Depot', 'VerticalAlignment', 'bottom', 'HorizontalAlignment', 'right');
    for i = 2:length(customers)
        plot(customers(i).id, 0, 'bo', 'MarkerSize', 8, 'LineWidth', 1);
    end
    colors = {'r', 'g', 'b', 'm', 'c'};
    for v = 1:NUM_VEHICLES
        route = solution{v};
        if ~isempty(route)
            x_coords = [customers(1).id, [customers(route).id]];
            y_coords = zeros(1, length(x_coords));
            plot(x_coords, y_coords, '-o', 'Color', colors{v}, 'LineWidth', 1.5, 'MarkerSize', 8);
        end
    end
    legend_entries = cell(1, NUM_VEHICLES);
    for v = 1:NUM_VEHICLES
        legend_entries{v} = sprintf('Vehicle %d', v);
    end
    legend(['Depot', 'Customers', legend_entries], 'Location', 'best');
    hold off;
end
best_solutions = cell(PERIOD_LENGTH, 1);
execution_times = zeros(PERIOD_LENGTH, 1);
total_distances = zeros(PERIOD_LENGTH, 1);
for period = 1:PERIOD_LENGTH
    tic;
    best_solutions{period} = simulated_annealing(customers, NUM_VEHICLES, MAX_ITER, INITIAL_TEMP, COOLING_RATE, VEHICLE_CAPACITY, calculate_distance, @calculate_route_distance);
    execution_times(period) = toc;
    fprintf('Best solution found for Day %d:\n', period);
    for v = 1:NUM_VEHICLES
        fprintf('Vehicle %d: ', v);
        fprintf('%d ', best_solutions{period}{v});
        fprintf('\n');
    end
    total_distance = 0;
    for v = 1:NUM_VEHICLES
        total_distance = total_distance + calculate_route_distance(best_solutions{period}{v}, customers, calculate_distance);
    end
    total_distances(period) = total_distance;
    fprintf('Total Distance for Day %d: %f\n', period, total_distance);
    fprintf('Execution Time for Day %d: %f seconds\n', period, execution_times(period));
    plot_routes(customers, best_solutions{period}, NUM_VEHICLES);
end

File: solutions/pvrp.cpp
// Period Vehicle Routing Problem (PVRP)
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>
#include <random>
#include <chrono>

using namespace std;

const int NUM_CUSTOMERS = 20;   
const int NUM_VEHICLES = 4;       
const int VEHICLE_CAPACITY = 10;  
const int MAX_ITER = 10000;       
const double INITIAL_TEMP = 100.0;
const double COOLING_RATE = 0.003;
const int PERIOD_LENGTH = 7;      
struct Customer {
    int id;
    int demand;
};
struct Depot {
    int id;
};
double distance(const Customer& c1, const Customer& c2) {
    return sqrt(pow(c1.id - c2.id, 2));
}

double calculate_route_distance(const vector<int>& route, const vector<Customer>& customers) {
    double total_distance = 0.0;
    for (int i = 0; i < route.size() - 1; ++i) {
        total_distance += distance(customers[route[i]], customers[route[i + 1]]);
    }
    total_distance += distance(customers[route.back()], customers[0]);
    return total_distance;
}

vector<vector<int>> generate_initial_solution(const vector<Customer>& customers) {
    vector<vector<int>> initial_solution(NUM_VEHICLES);

    int current_capacity = 0;
    int current_vehicle = 0;
    vector<bool> visited(customers.size(), false);
    visited[0] = true;
    int current_customer = 0; 
    while (true) {
        int nearest_customer = -1;
        double min_distance = numeric_limits<double>::max();
        for (int i = 1; i < customers.size(); ++i) {
            if (!visited[i] && customers[i].demand + current_capacity <= VEHICLE_CAPACITY) {
                double dist = distance(customers[current_customer], customers[i]);
                if (dist < min_distance) {
                    min_distance = dist;
                    nearest_customer = i;
                }
            }
        }
        if (nearest_customer == -1) {
            current_vehicle++;
            if (current_vehicle >= NUM_VEHICLES) {
                break; 
            }
            current_capacity = 0;
            current_customer = 0;
        } else {
           initial_solution[current_vehicle].push_back(nearest_customer);
            visited[nearest_customer] = true;
            current_capacity += customers[nearest_customer].demand;
            current_customer = nearest_customer;
        }
    }

    return initial_solution;
}
vector<vector<int>> simulated_annealing(const vector<Customer>& customers) {
    vector<vector<int>> current_solution = generate_initial_solution(customers);
    vector<vector<int>> best_solution = current_solution;
    double temperature = INITIAL_TEMP;
    random_device rd;
    mt19937 gen(rd());
    uniform_real_distribution<> dis(0.0, 1.0);
    for (int iter = 0; iter < MAX_ITER; ++iter) {
        vector<vector<int>> new_solution = current_solution;
        int vehicle1 = rand() % NUM_VEHICLES;
        int vehicle2 = rand() % NUM_VEHICLES;
        if (vehicle1 != vehicle2 && !new_solution[vehicle1].empty() && !new_solution[vehicle2].empty()) {
            int rand_index1 = rand() % new_solution[vehicle1].size();
            int rand_index2 = rand() % new_solution[vehicle2].size();
            swap(new_solution[vehicle1][rand_index1], new_solution[vehicle2][rand_index2]);
        }
        double current_cost = 0.0;
        double new_cost = 0.0;
        for (int v = 0; v < NUM_VEHICLES; ++v) {
            current_cost += calculate_route_distance(current_solution[v], customers);
            new_cost += calculate_route_distance(new_solution[v], customers);
        }
        if (new_cost < current_cost) {
            current_solution = new_solution;
        } else {
            double acceptance_probability = exp((current_cost - new_cost) / temperature);
            double random_value = dis(gen);
            if (random_value < acceptance_probability) {
                current_solution = new_solution;
            }
        }
        if (calculate_route_distance(current_solution[0], customers) < calculate_route_distance(best_solution[0], customers)) {
            best_solution = current_solution;
        }
        temperature *= (1 - COOLING_RATE);
    }
    return best_solution;
}

int main() {
    vector<Customer> customers(NUM_CUSTOMERS + 1);  
    for (int i = 1; i <= NUM_CUSTOMERS; ++i) {
        customers[i].id = i;
        customers[i].demand = rand() % 5 + 1;  
    }
    for (int period = 0; period < PERIOD_LENGTH; ++period) {
        auto start_time = chrono::high_resolution_clock::now();
        vector<vector<int>> solution = simulated_annealing(customers);
        auto end_time = chrono::high_resolution_clock::now();
        cout << "Best solution found for Day " << period + 1 << ":" << endl;
        for (int v = 0; v < NUM_VEHICLES; ++v) {
            cout << "Vehicle " << v + 1 << ": ";
            for (int i = 0; i < solution[v].size(); ++i) {
                cout << solution[v][i] << " ";
            }
            cout << endl;
        }
        double total_distance = 0.0;
        for (int v = 0; v < NUM_VEHICLES; ++v) {
            total_distance += calculate_route_distance(solution[v], customers);
        }
        cout << "Total Distance for Day " << period + 1 << ": " << total_distance << endl;
        chrono::duration<double> elapsed_time = end_time - start_time;
        cout << "Execution Time for Day " << period + 1 << ": " << elapsed_time.count() << " seconds" << endl;
    }
    return 0;
}
